File: elasticsearch-indexer/src/main/scala/co/elastic/codesearch/index/elasticsearch/ElasticsearchSourceIndexer.scala
/*
 * ELASTICSEARCH CONFIDENTIAL
 * __________________
 *
 *  Copyright Elasticsearch B.V. All rights reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Elasticsearch B.V. and its suppliers, if any.
 * The intellectual and technical concepts contained herein
 * are proprietary to Elasticsearch B.V. and its suppliers and
 * may be covered by U.S. and Foreign Patents, patents in
 * process, and are protected by trade secret or copyright
 * law.  Dissemination of this information or reproduction of
 * this material is strictly forbidden unless prior written
 * permission is obtained from Elasticsearch B.V.
 */

package co.elastic.codesearch.index.elasticsearch

import co.elastic.codesearch.index.SourceIndexer
import co.elastic.codesearch.model
import co.elastic.codesearch.model.{Language, LanguageElement}
import com.sksamuel.elastic4s.ElasticClient
import com.sksamuel.elastic4s.ElasticDsl.{properties, _}
import com.sksamuel.elastic4s.fields.{ElasticField, IntegerField, KeywordField, NestedField, ObjectField, TextField}
import com.typesafe.scalalogging.Logger

import scala.concurrent.{ExecutionContext, Future}

class ElasticsearchSourceIndexer(
  client: ElasticClient,
  codeSearchVersion: String,
  language: Language,
  languageElementTemplate: Seq[ElasticField],
  languageElementFields: LanguageElement => Map[String, Any],
  indexFmt: String = "code-search-%s-%s%s",
  templateFmt: String = "code-search-%s-%s%s"
)(implicit ec: ExecutionContext)
  extends SourceIndexer {

  val logger = Logger[ElasticsearchSourceIndexer]

  lazy val indexName =
    String
      .format(indexFmt, codeSearchVersion, language.name, language.version.map(v => s"-${v.value}").getOrElse(""))
      .toLowerCase
  lazy val templateName =
    String
      .format(
        templateFmt,
        codeSearchVersion,
        language.name,
        language.version.map(v => s"-${v.value}").getOrElse("")
      )
      .toLowerCase

  override def indexFile(file: model.SourceFile): Future[Either[SourceIndexer.IndexError, Unit]] = {
    logger.info(s"Indexing file [${file.fileName.value}]...")
    for {
      resp <- client.execute {
        indexInto(indexName).withId(file.path).fields(
          "version" -> file.version.value,
          "language" -> Map(
            "name" -> language.name,
            "version" -> language.version.map(_.value)
          ),
          "file_name" -> file.fileName.value,
          "path" -> file.path,
          "source" -> Map(
            "content" -> file.source
          ),
          "spans" -> file.spans.toList.map { span =>
            Map(
              "start" -> span.start,
              "end" -> span.end,
              "element" -> (Map(
                "kind" -> span.element.kind.value
              ) ++ languageElementFields(span.element))
            )
          }
        )
      }
      _ <- if (resp.status >= 400) {
        Future.failed(new Exception(s"Unexpected status [${resp.status}] from Elasticsearch: ${resp.error}"))
      } else {
        Future.successful(())
      }
      _ = logger.info(s"Indexed file [${file.fileName.value}]")
    } yield {
      Right(())
    }
  }

  override def setup(): Future[Either[SourceIndexer.SetupError, Unit]] = {
    logger.info(s"Setting up Elasticsearch template for Code Search ${codeSearchVersion}")
    for {
      resp <- client.execute {
        createIndexTemplate(templateName, List(indexName)).mappings(
          properties(
            KeywordField("version"),
            ObjectField(
              name = "language",
              properties = Seq(
                KeywordField("name"),
                KeywordField("version")
              )
            ),
            KeywordField("file_name"),
            KeywordField("path"),
            ObjectField(
              name = "source",
              enabled = Some(false),
              properties = Seq(
                TextField("contents")
              )
            ),
            NestedField(
              name = "spans",
              properties = Seq(
                IntegerField("start"),
                IntegerField("end"),
                ObjectField(
                  name = "element",
                  properties = (Seq(
                    KeywordField("kind")
                  ) ++ languageElementTemplate)
                )
              )
            )
          )
        )
      }
      _ <- if (resp.status >= 400) {
        Future.failed(new Exception(s"Unexpected status [${resp.status}] from Elasticsearch: ${resp.error}"))
      } else {
        Future.successful(())
      }
      _ = logger.info(s"Finished setting up Elasticsearch template for Code Search ${codeSearchVersion}")
    } yield {
      Right(())
    }
  }
}

object ElasticsearchSourceIndexer {
  sealed trait CreateTemplateError
}

File: java-compiler-plugin/src/main/java/co/elastic/codesearch/java/ElasticsearchSourceIndexer.java
package co.elastic.codesearch.java;

import co.elastic.clients.elasticsearch.ElasticsearchClient;
import co.elastic.clients.elasticsearch._types.mapping.KeywordProperty;
import co.elastic.clients.elasticsearch._types.mapping.ObjectProperty;
import co.elastic.clients.elasticsearch._types.mapping.Property;
import co.elastic.clients.elasticsearch.indices.PutIndexTemplateResponse;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;

public class ElasticsearchSourceIndexer {
    private final ElasticsearchClient client;
    private final ExecutorService executorService;
    private String javaVersion;
    private String codeSearchVersion;
    private final String indexFmt;
    private final String templateFmt;

    public ElasticsearchSourceIndexer(ElasticsearchClient client, ExecutorService executorService, String javaVersion, String codeSearchVersion, String indexFmt, String templateFmt) {
        this.client = client;
        this.executorService = executorService;
        this.javaVersion = javaVersion;
        this.codeSearchVersion = codeSearchVersion;
        this.indexFmt = indexFmt;
        this.templateFmt = templateFmt;
    }

    public Future<Void> indexFile(SourceFile file) {
//        return executorService.submit(new Callable<Void>() {
//            @Override
//            public Void call() throws Exception {
//                client.index(i -> i.index(getIndexName()).)
//            }
//        })
        return null;
    }

    private String getTemplateName() {
        return String.format(templateFmt, codeSearchVersion, "java", javaVersion);
    }

    private String getIndexName() {
        return String.format(indexFmt, codeSearchVersion, "java", javaVersion);
    }

    public Future<Void> setup() {
        return executorService.submit(new Callable<Void>() {
            @Override
            public Void call() throws Exception {
                client.indices().putIndexTemplate(i ->
                        i.name(getTemplateName()).indexPatterns(getIndexName()).template(t ->
                                t.mappings(m ->
                                        m.properties("version", p -> p.keyword(KeywordProperty.of(k -> k)))
                                                .properties("language", p -> p.object(ObjectProperty.of(o ->
                                                        o.properties("name", pp -> pp.keyword(KeywordProperty.of(k -> k)))
                                                                .properties("version", pp -> pp.keyword(KeywordProperty.of(k -> k))))))
                                                .properties("file_name", p -> p.keyword(k -> k))
                                                .properties("path", p -> p.keyword(k -> k))
                                                .properties("source", p -> p.object(o ->
                                                        o.enabled(false).properties("contents", pp -> pp.text(tt -> tt))))
                                                .properties("spans", p -> p.nested(n ->
                                                        n.properties("start", pp -> pp.integer(ii -> ii))
                                                                .properties("end", pp -> pp.integer(ii -> ii))
                                                                .properties("element", pp -> pp.object(o ->
                                                                        o.properties("kind", ppp -> ppp.keyword(k -> k))
                                                                ))
                                                ))
                                )
                        )
                );
                return null;
            }
        });
    }
}
